<a href="https://colab.research.google.com/github/Baptiste-Boursin/Baptiste-Boursin/blob/main/Copie_de_tp1_prepa_feature.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Import des librairies ##

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
#Imports librairies de base
import numpy as np
import pandas as pd

#Bibliothèques pour Cross validation
import random
import re

#Affichage de Graphes
from IPython.display import Markdown, display, HTML

## 1. Lecture des données ##

1.1 Charger les données d'apprentissage (Data/passagers.csv) dans un DataFrame nommé train. Afficher 10 exemples.

In [3]:
train = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Data/passagers.csv', dtype={'PassengerId': np.int})
train.sample(5)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  """Entry point for launching an IPython kernel.


PassengerId  Survived  Pclass                        Name   Sex   Age  \
492          493         0       1  Molson, Mr. Harry Markland  male  55.0   
531          532         0       3           Toufik, Mr. Nakli  male   NaN   
196          197         0       3         Mernagh, Mr. Robert  male   NaN   
762          763         1       3       Barah, Mr. Hanna Assi  male  20.0   
74            75         1       3               Bing, Mr. Lee  male  32.0   

     SibSp  Parch  Ticket     Fare Cabin Embarked  
492      0      0  113787  30.5000   C30        S  
531      0      0    2641   7.2292   NaN        C  
196      0      0  368703   7.7500   NaN        Q  
762      0      0    2663   7.2292   NaN        C  
74       0      0    1601  56.4958   NaN        S

1.2 Charger les données d'apprentissage (Data/test.csv) dans un Dataframe nommé test. Afficher 10 exemples. 

In [4]:
test = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Data/test.csv', dtype={'PassengerId': np.int})

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  """Entry point for launching an IPython kernel.


In [5]:
test.sample(5)

PassengerId  Pclass                                        Name     Sex  \
230         1122       2                 Sweet, Mr. George Frederick    male   
174         1066       3  Asplund, Mr. Carl Oscar Vilhelm Gustafsson    male   
416         1308       3                         Ware, Mr. Frederick    male   
37           929       3                          Cacic, Miss. Manda  female   
170         1062       3                          Lithman, Mr. Simon    male   

      Age  SibSp  Parch         Ticket     Fare Cabin Embarked  
230  14.0      0      0         220845  65.0000   NaN        S  
174  40.0      1      5         347077  31.3875   NaN        S  
416   NaN      0      0         359309   8.0500   NaN        S  
37   21.0      0      0         315087   8.6625   NaN        S  
170   NaN      0      0  S.O./P.P. 251   7.5500   NaN        S

1.3 Afficher les informations des deux DataFrames train et Test.

In [6]:
display(train)
display(test)

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292   NaN        Q  
1    female  47.0      1      0              363272    7.0000   NaN        S  
2      male  62.0      0      0              240276    9.6875   NaN        Q  
3      male  27.0      0      0              315154    8.6625   NaN        S  
4    female  22.0      1      1             3101298   12.2875   NaN        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male   NaN      0      0              359309    8.0500   NaN        S  
417    male   NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 11 columns]

1.4 Calculer le taux global de survie

In [7]:
train['Survived'].mean()

0.3838383838383838

#### Question 1 : Quel Travail voyez-vous pour que les données soient prêtes à faire tourner dans un modèle ?




#### Question 2 : Quelle différence y-a-t'il entre les données train et les données test ? 
Les données train comprennent les données des passagers 1 à 891 en indiquant s'ils ont survécu où non.
Les données 'test' représentent les passagers 892 à 1309 mais le colonne 'Survived' qui indique si le passager à survécu ou non n'est pas affichée.



#### Question 3 : Quel est le rôle de ces deux sets de données ?
Le rôle de 'train' est d'entrainer notre machine pour qu'il puisse ensuite indiquer pour les données 'test' si les utilisateurs on survécu où non.



## 2. Analyse des features

2.1 Analyser l'effet de la variable Classe sur le taux de survie, en affichant le % de survie par classe

In [8]:
train[['Survived','Pclass']].groupby(['Pclass']).mean()*100

Survived
Pclass           
1       62.962963
2       47.282609
3       24.236253

2.2 Analyser l'effet de la variable Genre (Sex) sur le taux de survie, en affichant le % de survie par genre

In [9]:
train[['Survived','Sex']].groupby(['Sex']).mean()*100

Survived
Sex              
female  74.203822
male    18.890815

2.3 Calculer une nouvelle colonne FamilySize contenant la taille de la famille, pour les DataFrame train et Test. Afficher pour 10 exemples l'Id du passager, le nom, l'age, le sexe et la taille de la famille.

In [10]:
train['FamilySize'] = train['Parch']+train['SibSp']+1
test['FamilySize'] = test['Parch']+test['SibSp']+1
display(train[['PassengerId','Name','Age','Sex','FamilySize']].sample(10))
display(test[['PassengerId','Name','Age','Sex','FamilySize']].sample(10))

PassengerId                                               Name   Age  \
634          635                                 Skoog, Miss. Mabel   9.0   
518          519  Angle, Mrs. William A (Florence "Mary" Agnes H...  36.0   
655          656                          Hickman, Mr. Leonard Mark  24.0   
669          670  Taylor, Mrs. Elmer Zebley (Juliet Cummins Wright)   NaN   
144          145                         Andrew, Mr. Edgardo Samuel  18.0   
746          747                        Abbott, Mr. Rossmore Edward  16.0   
650          651                                  Mitkoff, Mr. Mito   NaN   
127          128                          Madsen, Mr. Fridtjof Arne  24.0   
0              1                            Braund, Mr. Owen Harris  22.0   
658          659                       Eitemiller, Mr. George Floyd  23.0   

        Sex  FamilySize  
634  female           6  
518  female           2  
655    male           3  
669  female           2  
144    male           1  
746    male           3  
650    male           1  
127    male           1  
0      male           2  
658    male           1

PassengerId                                              Name   Age  \
303         1195                                 Pokrnic, Mr. Tome  24.0   
312         1204                               Sadowitz, Mr. Harry   NaN   
155         1047                              Duquemin, Mr. Joseph  24.0   
192         1084                 van Billiard, Master. Walter John  11.5   
279         1171                         Oxenham, Mr. Percy Thomas  22.0   
225         1117  Moubarek, Mrs. George (Omine Amenia" Alexander)"   NaN   
156         1048                                 Bird, Miss. Ellen  29.0   
298         1190                        Loring, Mr. Joseph Holland  30.0   
172         1064                          Dyker, Mr. Adolf Fredrik  23.0   
337         1229                                 Elias, Mr. Joseph  39.0   

        Sex  FamilySize  
303    male           1  
312    male           1  
155    male           1  
192    male           3  
279    male           1  
225  female           3  
156  female           1  
298    male           1  
172    male           2  
337    male           3

2.4 Visualisez le nombre de personnes et le taux de survie par taille de famille.

In [11]:
train[['Survived','FamilySize']].groupby(['FamilySize']).agg(['count', 'mean'])


Survived          
              count      mean
FamilySize                   
1               537  0.303538
2               161  0.552795
3               102  0.578431
4                29  0.724138
5                15  0.200000
6                22  0.136364
7                12  0.333333
8                 6  0.000000
11                7  0.000000

2.5 Calculer une nouvelle colonne IsAlone indiquant si le passager est seul ou pas, pour les DataFrame train et Test. Afficher pour 10 exemples l'Id du passager, le nom, l'age, le sexe et le status IsAlone.

In [12]:
train['isAlone'] = train['FamilySize']==1
test['isAlone'] = test['FamilySize']==1
display(train[['PassengerId','Name','Age','Sex','isAlone']].sample(10))
display(test[['PassengerId','Name','Age','Sex','isAlone']].sample(10))

PassengerId                                 Name   Age     Sex  isAlone
188          189                     Bourke, Mr. John  40.0    male    False
717          718  Troutt, Miss. Edwina Celia "Winnie"  27.0  female     True
396          397                  Olsson, Miss. Elina  31.0  female     True
118          119             Baxter, Mr. Quigg Edmond  24.0    male    False
619          620                  Gavey, Mr. Lawrence  26.0    male     True
596          597           Leitch, Miss. Jessie Wills   NaN  female     True
428          429                     Flynn, Mr. James   NaN    male     True
476          477              Renouf, Mr. Peter Henry  34.0    male    False
585          586                  Taussig, Miss. Ruth  18.0  female    False
147          148     Ford, Miss. Robina Maggie "Ruby"   9.0  female    False

PassengerId                                              Name   Age  \
224         1116  Candee, Mrs. Edward (Helen Churchill Hungerford)  53.0   
5            897                        Svensson, Mr. Johan Cervin  14.0   
160         1052                                Smyth, Miss. Julia   NaN   
389         1281                       Palsson, Master. Paul Folke   6.0   
203         1095                        Quick, Miss. Winifred Vera   8.0   
149         1041                            Lahtinen, Rev. William  30.0   
396         1288                              Colbert, Mr. Patrick  24.0   
192         1084                 van Billiard, Master. Walter John  11.5   
348         1240                                  Giles, Mr. Ralph  24.0   
148         1040                         Crafton, Mr. John Bertram   NaN   

        Sex  isAlone  
224  female     True  
5      male     True  
160  female     True  
389    male    False  
203  female    False  
149    male    False  
396    male     True  
192    male    False  
348    male     True  
148    male     True

2.6 Visualisez le nombre de personnes et le taux de survie par statut seul ou pas.

In [13]:
train[['Survived','isAlone']].groupby(['isAlone']).agg(['count', 'mean'])

Survived          
           count      mean
isAlone                   
False        354  0.505650
True         537  0.303538

### Question 1 : La variable Classe est-elle utile pour prévoir la survie ? Pourquoi ?
Oui elle est utile, car plus la clase est haute plus le nombre de survivant augmente. Les personnes qui étaient en première classe avaient plus de chances de survies.



### Question 2 : Même question pour la variable Sex 
Oui, il y a une grande différence de survie entre les hommes te les femmes.

### Question 3 : Comment traiter le variables liées à la famille du passager ?





## 3. Traitement des données manquantes (à faire sur les données Train et Test)

3.1 La variable "Embarked" possède des valeurs nulles, compléter par la valeur majoritaire (S).

In [14]:
train['Embarked'].fillna('S',inplace=True)
test['Embarked'].fillna('S',inplace=True)

3.2 Afficher l'influence du port d'embarquement sur la Survie 

In [15]:
train[['Survived','Embarked']].groupby(['Embarked']).mean()*100

Survived
Embarked           
C         55.357143
Q         38.961039
S         33.900929

3.3 La variable "Fare" (prix du ticket) possède des valeurs manquantes. Remplacer ces valeurs par la valeur médiane.

In [16]:
train['Fare'].fillna(train['Fare'].median(),inplace=True)
test['Fare'].fillna(test['Fare'].median(),inplace=True)

3.4 Créer une variable CategoricalFare Découpant la valeur de ticket en 4 catégories possédant des nombres de passagers à peu près égaux. 

In [17]:
train['CategoricalFare'] = pd.qcut(train['Fare'],4) 
test['CategoricalFare'] = pd.qcut(test['Fare'],4) 


3.5 Analyser l'effet de ces catégories de prix de ticket sur la survie.

In [18]:
train[['Survived','CategoricalFare']].groupby(['CategoricalFare']).mean() * 100

Survived
CategoricalFare           
(-0.001, 7.91]   19.730942
(7.91, 14.454]   30.357143
(14.454, 31.0]   45.495495
(31.0, 512.329]  58.108108

3.6 Combien de valeurs d'Age manquent ? Complète par des valeurs entières aléatoires entre (moyenne - écart type) et (moyenne + écart type).

In [19]:
#train['Age'].isna().sum() #= 177
train['Age'].fillna(np.random.randint(train['Age'].mean() - train['Age'].std(),train['Age'].mean() + train['Age'].std()),inplace=True)
test['Age'].fillna(np.random.randint(test['Age'].mean() - test['Age'].std(),test['Age'].mean() + test['Age'].std()),inplace=True)


3.7 Créer une variable CategoricalAge Découpant la valeur d'age en 5 catégories égales. 

In [20]:
train['CategoricalAge'] = pd.qcut(train['Age'],5) 
test['CategoricalAge'] = pd.qcut(test['Age'],5)
train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  20.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  FamilySize  isAlone  \
0        0         A/5 21171   7.2500   NaN        S           2    False   
1        0          PC 17599  71.2833   C85        C           2    False   
2        0  STON/O2. 3101282   7.9250   NaN        S           1     True   
3        0            113803  53.1000  C123        S           2    False   
4        0            373450   8.0500   NaN        S           1     True   
..     ...               ...      ...   ...      ...         ...      ...   
886      0            211536  13.0000   NaN        S           1     True   
887      0            112053  30.0000   B42        S           1     True   
888      2        W./C. 6607  23.4500   NaN        S           4    False   
889      0            111369  30.0000  C148        C           1     True   
890      0            370376   7.7500   NaN        Q           1     True   

     CategoricalFare CategoricalAge  
0     (-0.001, 7.91]   (20.5, 28.0]  
1    (31.0, 512.329]   (28.0, 38.0]  
2     (7.91, 14.454]   (20.5, 28.0]  
3    (31.0, 512.329]   (28.0, 38.0]  
4     (7.91, 14.454]   (28.0, 38.0]  
..               ...            ...  
886   (7.91, 14.454]   (20.5, 28.0]  
887   (14.454, 31.0]  (0.419, 20.0]  
888   (14.454, 31.0]  (0.419, 20.0]  
889   (14.454, 31.0]   (20.5, 28.0]  
890   (-0.001, 7.91]   (28.0, 38.0]  

[891 rows x 16 columns]

3.8 Analyser l'effet de ces catégories d'age sur la survie.

In [21]:
train[['Survived','CategoricalAge']].groupby(['CategoricalAge']).mean() * 100

Survived
CategoricalAge           
(0.419, 20.0]   37.640449
(20.0, 20.5]     0.000000
(20.5, 28.0]    36.263736
(28.0, 38.0]    43.428571
(38.0, 80.0]    37.288136

### Question 1 : Classer les variables par ordre d'intérêt

## 4 : Traitement de features complexes

4.1 : Le nom permet de retrouver le titre du passager. Trouver les titres lorsqu'ils existent.

In [22]:
train['Title']=train.Name.apply(lambda st: st[st.find(" ")+1:st.find(".")])
test['Title'] = test.Name.apply(lambda st: st[st.find(" ")+1:st.find(".")])
test

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  \
0      male  34.5      0      0              330911    7.8292   NaN        Q   
1    female  47.0      1      0              363272    7.0000   NaN        S   
2      male  62.0      0      0              240276    9.6875   NaN        Q   
3      male  27.0      0      0              315154    8.6625   NaN        S   
4    female  22.0      1      1             3101298   12.2875   NaN        S   
..      ...   ...    ...    ...                 ...       ...   ...      ...   
413    male  20.0      0      0           A.5. 3236    8.0500   NaN        S   
414  female  39.0      0      0            PC 17758  108.9000  C105        C   
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S   
416    male  20.0      0      0              359309    8.0500   NaN        S   
417    male  20.0      1      1                2668   22.3583   NaN        C   

     FamilySize  isAlone    CategoricalFare CategoricalAge          Title  
0             1     True    (-0.001, 7.896]   (27.0, 39.0]             Mr  
1             2    False    (-0.001, 7.896]   (39.0, 76.0]            Mrs  
2             1     True    (7.896, 14.454]   (39.0, 76.0]             Mr  
3             1     True    (7.896, 14.454]   (21.0, 27.0]             Mr  
4             3    False    (7.896, 14.454]   (21.0, 27.0]            Mrs  
..          ...      ...                ...            ...            ...  
413           1     True    (7.896, 14.454]  (0.169, 20.0]             Mr  
414           1     True  (31.472, 512.329]   (27.0, 39.0]  y Ocana, Dona  
415           1     True    (-0.001, 7.896]   (27.0, 39.0]             Mr  
416           1     True    (7.896, 14.454]  (0.169, 20.0]             Mr  
417           3    False   (14.454, 31.472]  (0.169, 20.0]         Master  

[418 rows x 16 columns]

4.2 : En utilisant la fonction pandas crosstab, afficher pour train le nombre d'occurences par titres (en ligne) et par sexe (en colonne)

In [23]:
pd.crosstab(train.Title,train.Sex)

Sex                female  male
Title                          
Billiard, Mr            0     1
Capt                    0     1
Carlo, Mr               0     1
Col                     0     2
Cruyssen, Mr            0     1
Don                     0     1
Dr                      1     6
Gordon, Lady            1     0
Gordon, Sir             0     1
Impe, Miss              1     0
Impe, Mr                0     1
Impe, Mrs               1     0
Jonkheer                0     1
Major                   0     2
Master                  0    40
Melkebeke, Mr           0     1
Messemaeker, Mrs        1     0
Miss                  179     0
Mlle                    2     0
Mme                     1     0
Mr                      0   502
Mrs                   121     0
Ms                      1     0
Mulder, Mr              0     1
Pelsmaeker, Mr          0     1
Planke, Miss            1     0
Planke, Mr              0     1
Planke, Mrs             1     0
Rev                     0     6
Shawah, Mr              0     1
Steen, Mr               0     1
Velde, Mr               0     1
Walle, Mr               0     1
der hoef, Mr            0     1
the Countess            1     0
y Castellana, Mr        0     1
y Castellana, Mrs       1     0
y Manent, Mr            0     1
y More, Miss            1     0

4.3 : 
Remplacer 'Melle et 'Ms' par 'Miss'
Remplacer 'Mme' par 'Mrs'
Creer une valeur "Rare" pour les autres titres avec moins de 10 occurences. 

In [24]:
train['Title'] = train['Title'].str.replace('Mlle|Ms','Miss')
train['Title'] = train['Title'].str.replace('Mme','Mrs')
test['Title'] = test['Title'].str.replace('Mlle|Ms','Miss')
test['Title'] = test['Title'].str.replace('Mme','Mrs')

valuesTrain = train.Title.value_counts()
toChangeTrain = train.Title.isin(valuesTrain.index[valuesTrain<10])
train.loc[toChangeTrain, 'Title'] = 'Rare'

valuesTest = test.Title.value_counts()
toChangeTest = test.Title.isin(valuesTest.index[valuesTest<10])
test.loc[toChangeTest, 'Title'] = 'Rare'
test

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  This is separate from the ipykernel package so we can avoid doing imports until


PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  \
0      male  34.5      0      0              330911    7.8292   NaN        Q   
1    female  47.0      1      0              363272    7.0000   NaN        S   
2      male  62.0      0      0              240276    9.6875   NaN        Q   
3      male  27.0      0      0              315154    8.6625   NaN        S   
4    female  22.0      1      1             3101298   12.2875   NaN        S   
..      ...   ...    ...    ...                 ...       ...   ...      ...   
413    male  20.0      0      0           A.5. 3236    8.0500   NaN        S   
414  female  39.0      0      0            PC 17758  108.9000  C105        C   
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S   
416    male  20.0      0      0              359309    8.0500   NaN        S   
417    male  20.0      1      1                2668   22.3583   NaN        C   

     FamilySize  isAlone    CategoricalFare CategoricalAge   Title  
0             1     True    (-0.001, 7.896]   (27.0, 39.0]      Mr  
1             2    False    (-0.001, 7.896]   (39.0, 76.0]     Mrs  
2             1     True    (7.896, 14.454]   (39.0, 76.0]      Mr  
3             1     True    (7.896, 14.454]   (21.0, 27.0]      Mr  
4             3    False    (7.896, 14.454]   (21.0, 27.0]     Mrs  
..          ...      ...                ...            ...     ...  
413           1     True    (7.896, 14.454]  (0.169, 20.0]      Mr  
414           1     True  (31.472, 512.329]   (27.0, 39.0]    Rare  
415           1     True    (-0.001, 7.896]   (27.0, 39.0]      Mr  
416           1     True    (7.896, 14.454]  (0.169, 20.0]      Mr  
417           3    False   (14.454, 31.472]  (0.169, 20.0]  Master  

[418 rows x 16 columns]

## 5. Mise en forme des données non numériques

5.1 Attribuer une valeur numérique au Genre. Par exemple 0 pour 'female' et 1 pour 'male'

In [25]:
train['Sex'] =  train['Sex'].map(
    {'female':0,
     'male':1
    }
)


test['Sex'] =  test['Sex'].map(
    {'female':0,
     'male':1
    }
)
test

PassengerId  Pclass                                          Name  Sex  \
0            892       3                              Kelly, Mr. James    1   
1            893       3              Wilkes, Mrs. James (Ellen Needs)    0   
2            894       2                     Myles, Mr. Thomas Francis    1   
3            895       3                              Wirz, Mr. Albert    1   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)    0   
..           ...     ...                                           ...  ...   
413         1305       3                            Spector, Mr. Woolf    1   
414         1306       1                  Oliva y Ocana, Dona. Fermina    0   
415         1307       3                  Saether, Mr. Simon Sivertsen    1   
416         1308       3                           Ware, Mr. Frederick    1   
417         1309       3                      Peter, Master. Michael J    1   

      Age  SibSp  Parch              Ticket      Fare Cabin Embarked  \
0    34.5      0      0              330911    7.8292   NaN        Q   
1    47.0      1      0              363272    7.0000   NaN        S   
2    62.0      0      0              240276    9.6875   NaN        Q   
3    27.0      0      0              315154    8.6625   NaN        S   
4    22.0      1      1             3101298   12.2875   NaN        S   
..    ...    ...    ...                 ...       ...   ...      ...   
413  20.0      0      0           A.5. 3236    8.0500   NaN        S   
414  39.0      0      0            PC 17758  108.9000  C105        C   
415  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S   
416  20.0      0      0              359309    8.0500   NaN        S   
417  20.0      1      1                2668   22.3583   NaN        C   

     FamilySize  isAlone    CategoricalFare CategoricalAge   Title  
0             1     True    (-0.001, 7.896]   (27.0, 39.0]      Mr  
1             2    False    (-0.001, 7.896]   (39.0, 76.0]     Mrs  
2             1     True    (7.896, 14.454]   (39.0, 76.0]      Mr  
3             1     True    (7.896, 14.454]   (21.0, 27.0]      Mr  
4             3    False    (7.896, 14.454]   (21.0, 27.0]     Mrs  
..          ...      ...                ...            ...     ...  
413           1     True    (7.896, 14.454]  (0.169, 20.0]      Mr  
414           1     True  (31.472, 512.329]   (27.0, 39.0]    Rare  
415           1     True    (-0.001, 7.896]   (27.0, 39.0]      Mr  
416           1     True    (7.896, 14.454]  (0.169, 20.0]      Mr  
417           3    False   (14.454, 31.472]  (0.169, 20.0]  Master  

[418 rows x 16 columns]

5.2 : Attribuer une valeur pour les possibilités de la variable "Title" ('Master','Miss','Mr','Mrs','Rare'). 

In [27]:
train['Title'] =  train['Title'].map(
    {'Master':0,
     'Miss':1,
     'Mr':2,
     'Mrs':3,
     'Rare':4
    }
)
train

test['Title'] =  test['Title'].map(
    {'Master':0,
     'Miss':1,
     'Mr':2,
     'Mrs':3,
     'Rare':4
    }
)
test

PassengerId  Pclass                                          Name  Sex  \
0            892       3                              Kelly, Mr. James    1   
1            893       3              Wilkes, Mrs. James (Ellen Needs)    0   
2            894       2                     Myles, Mr. Thomas Francis    1   
3            895       3                              Wirz, Mr. Albert    1   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)    0   
..           ...     ...                                           ...  ...   
413         1305       3                            Spector, Mr. Woolf    1   
414         1306       1                  Oliva y Ocana, Dona. Fermina    0   
415         1307       3                  Saether, Mr. Simon Sivertsen    1   
416         1308       3                           Ware, Mr. Frederick    1   
417         1309       3                      Peter, Master. Michael J    1   

      Age  SibSp  Parch              Ticket      Fare Cabin Embarked  \
0    34.5      0      0              330911    7.8292   NaN        Q   
1    47.0      1      0              363272    7.0000   NaN        S   
2    62.0      0      0              240276    9.6875   NaN        Q   
3    27.0      0      0              315154    8.6625   NaN        S   
4    22.0      1      1             3101298   12.2875   NaN        S   
..    ...    ...    ...                 ...       ...   ...      ...   
413  20.0      0      0           A.5. 3236    8.0500   NaN        S   
414  39.0      0      0            PC 17758  108.9000  C105        C   
415  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S   
416  20.0      0      0              359309    8.0500   NaN        S   
417  20.0      1      1                2668   22.3583   NaN        C   

     FamilySize  isAlone    CategoricalFare CategoricalAge  Title  
0             1     True    (-0.001, 7.896]   (27.0, 39.0]      2  
1             2    False    (-0.001, 7.896]   (39.0, 76.0]      3  
2             1     True    (7.896, 14.454]   (39.0, 76.0]      2  
3             1     True    (7.896, 14.454]   (21.0, 27.0]      2  
4             3    False    (7.896, 14.454]   (21.0, 27.0]      3  
..          ...      ...                ...            ...    ...  
413           1     True    (7.896, 14.454]  (0.169, 20.0]      2  
414           1     True  (31.472, 512.329]   (27.0, 39.0]      4  
415           1     True    (-0.001, 7.896]   (27.0, 39.0]      2  
416           1     True    (7.896, 14.454]  (0.169, 20.0]      2  
417           3    False   (14.454, 31.472]  (0.169, 20.0]      0  

[418 rows x 16 columns]

5.3 Remplacer les titres manquant par 0, et attribuer à la colonne le type entier.

In [28]:
train['Title'].fillna(5,inplace=True)
train['Title'].astype('int32').dtypes
test['Title'].fillna(5,inplace=True)
test['Title'].astype('int32').dtypes

dtype('int32')

5.4 Attribuer une valeur numérique aux 4 catégories de la variable Fare (question 3.6)

In [29]:
# train['CategoricalFare'] =  train['CategoricalFare'].map(
# {'(-0.001, 7.91]' : 0,
# '(7.91, 14.454]' : 1,
# '(14.454, 31.0]' : 2,
# '(31.0, 512.329]' : 3}
# )
train['CategoricalFare'] =  train['CategoricalFare'].map(
    {
train['CategoricalFare'].loc[0] : 0,
train['CategoricalFare'].loc[1] : 1,
train['CategoricalFare'].loc[2] : 2,
train['CategoricalFare'].loc[3] : 3 
    })
train





PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name  Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    1  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...    0  38.0      1   
2                               Heikkinen, Miss. Laina    0  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)    0  35.0      1   
4                             Allen, Mr. William Henry    1  35.0      0   
..                                                 ...  ...   ...    ...   
886                              Montvila, Rev. Juozas    1  27.0      0   
887                       Graham, Miss. Margaret Edith    0  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"    0  20.0      1   
889                              Behr, Mr. Karl Howell    1  26.0      0   
890                                Dooley, Mr. Patrick    1  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  FamilySize  isAlone  \
0        0         A/5 21171   7.2500   NaN        S           2    False   
1        0          PC 17599  71.2833   C85        C           2    False   
2        0  STON/O2. 3101282   7.9250   NaN        S           1     True   
3        0            113803  53.1000  C123        S           2    False   
4        0            373450   8.0500   NaN        S           1     True   
..     ...               ...      ...   ...      ...         ...      ...   
886      0            211536  13.0000   NaN        S           1     True   
887      0            112053  30.0000   B42        S           1     True   
888      2        W./C. 6607  23.4500   NaN        S           4    False   
889      0            111369  30.0000  C148        C           1     True   
890      0            370376   7.7500   NaN        Q           1     True   

     CategoricalFare CategoricalAge  Title  
0                0.0   (20.5, 28.0]      2  
1                3.0   (28.0, 38.0]      3  
2                2.0   (20.5, 28.0]      1  
3                3.0   (28.0, 38.0]      3  
4                2.0   (28.0, 38.0]      2  
..               ...            ...    ...  
886              2.0   (20.5, 28.0]      4  
887              NaN  (0.419, 20.0]      1  
888              NaN  (0.419, 20.0]      1  
889              NaN   (20.5, 28.0]      2  
890              0.0   (28.0, 38.0]      2  

[891 rows x 17 columns]

5.5 Attribuer une valeur numérique aux 5 catégories de la variable Age (question 3.7)

5.6 Attibuer une valeur numérique au Port d'embarquement (Embarked)

5.7 Supprimer les features inutiles

5.8 Afficher les infos pour les deux datasets

# Sauver les deux datasets dans des fichier csv. Eviter d'écrire un index (n° de ligne)

In [ ]:
train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                             Montvila, Rare. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

##### Question : Sur combien de questions avez-vous utilisé les conseils ?
